qualtran/bloqs/block_encoding/tensor_product.py

```diff
@@ -0,0 +1,192 @@
+from qualtran.resource_counting import BloqCountT, SympySymbolAllocator
+from qualtran.symbolics import is_symbolic, prod, ssum, SymbolicFloat, SymbolicInt
+
+
+@frozen
+class TensorProduct(BlockEncoding):
+    r"""Tensor product of a sequence of block encodings.
+
+    Builds the block encoding as
+    $$
+        B[U_1 ⊗ U_2 ⊗ \cdots ⊗ U_n] = B[U_1] ⊗ B[U_2] ⊗ \cdots ⊗ B[U_n]
+    $$
+
+    When each $B[U_i]$ is a $(\alpha_i, a_i, \epsilon_i)$-block encoding of $U_i$, we have that
+    $B[U_1 ⊗ \cdots ⊗ U_n]$ is a $(\prod_i \alpha_i, \sum_i a_i, \sum_i \alpha_i \epsilon_i)$-block
+    encoding of $U_1 ⊗ \cdots ⊗ U_n$.
+
+    Args:
+        U: A sequence of block encodings.
+
+    Registers:
+        system: The system register.
+        ancilla: The ancilla register.
+        resource: The resource register.
+
+    References:
+        [Quantum algorithms: A survey of applications and end-to-end complexities](https://arxiv.org/abs/2310.03011). Dalzell et al. (2023). Ch. 10.2.
+    """
+
+    block_encodings: Tuple[BlockEncoding, ...] = field(
+        converter=lambda x: x if isinstance(x, tuple) else tuple(x), validator=validators.min_len(1)
+    )
+
+    @cached_property
+    def signature(self) -> Signature:
+        return Signature.build_from_dtypes(
+            system=QAny(self.system_bitsize),
+            ancilla=QAny(self.ancilla_bitsize),
+            resource=QAny(self.resource_bitsize),
+        )
+
+    @cached_property
+    def system_bitsize(self) -> SymbolicInt:
+        return ssum(u.system_bitsize for u in self.block_encodings)
+
+    def pretty_name(self) -> str:
+        return f"B[{'⊗'.join(u.pretty_name()[2:-1] for u in self.block_encodings)}]"
+
+    @cached_property
+    def alpha(self) -> SymbolicFloat:
+        return prod(u.alpha for u in self.block_encodings)
+
+    @cached_property
+    def ancilla_bitsize(self) -> SymbolicInt:
+        return ssum(u.ancilla_bitsize for u in self.block_encodings)
+
+    @cached_property
+    def resource_bitsize(self) -> SymbolicInt:
+        return ssum(u.resource_bitsize for u in self.block_encodings)
+
+    @cached_property
+    def epsilon(self) -> SymbolicFloat:
+        return ssum(u.alpha * u.epsilon for u in self.block_encodings)
+
+    @property
+    def target_registers(self) -> Tuple[Register, ...]:
+        return (self.signature.get_right("system"),)
+
+    @property
+    def junk_registers(self) -> Tuple[Register, ...]:
+        return (self.signature.get_right("resource"),)
+
+    @property
+    def selection_registers(self) -> Tuple[Register, ...]:
+        return (self.signature.get_right("ancilla"),)
+
+    @property
+    def signal_state(self) -> PrepareOracle:
+        """This method will be implemented in the future after PrepareOracle is updated for the BlockEncoding interface."""
+        raise NotImplementedError
+
+    def build_call_graph(self, ssa: SympySymbolAllocator) -> Set[BloqCountT]:
+        return {(bloq, 1) for bloq in self.block_encodings}
```

Comment:
Let's instead use a dictionary to add up the bloq counts and then convert to a set; so that if `self.block_encodings` has two entries for the same `bloq` then we have the right counts.